### PR TITLE
Add path viz & lineage example

### PR DESCRIPTION
**Pull request by skrawcz**
Adds path visualization and refactors the internal graphviz function a little.

## Changes
 - driver
 - graph and test
 - examples & docs

## How I tested this
 - locally
 - unit tests for existing code

## Notes
 - see image files for example

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/driver.py [hunk 1]
```diff
@@ -542,6 +545,114 @@ def what_is_upstream_of(self, *node_names: str) -> List[Variable]:
...
+        Assumes that the nodes exist in the graph.
+
+        :param upstream_node_name: the name of the node that we want to start from.
+        :param downstream_node_name: the name of the node that we want to end at.
+        :return: set of nodes that comprise the path between the two nodes, inclusive of the two nodes.
+        """
+        downstream_nodes = self.graph.get_impacted_nodes([upstream_node_name])
+        # we skip user_nodes because it'll be the upstream node, or it wont matter.
+        upstream_nodes, _ = self.graph.get_upstream_nodes([downstream_node_name])
+        nodes_for_path = set(downstream_nodes).intersection(set(upstream_nodes))
```
Comment (elijahbenizzy) [inline hamilton/driver.py:585]: Well this is quite clever

File: hamilton/driver.py [hunk 2]
```diff
@@ -542,6 +545,114 @@ def what_is_upstream_of(self, *node_names: str) -> List[Variable]:
...
+    ) -> Set[node.Node]:
+        """Gets the nodes representing the path between two nodes, inclusive of the two nodes.
+
+        Assumes that the nodes exist in the graph.
+
+        :param upstream_node_name: the name of the node that we want to start from.
+        :param downstream_node_name: the name of the node that we want to end at.
+        :return: set of nodes that comprise the path between the two nodes, inclusive of the two nodes.
+        """
+        downstream_nodes = self.graph.get_impacted_nodes([upstream_node_name])
```
Comment (elijahbenizzy) [inline hamilton/driver.py:582]: We should maybe call this `get_downstream_nodes` for consistency...

File: hamilton/graph.py [hunk 3]
```diff
@@ -95,43 +95,51 @@ def create_function_graph(
...
-        # checks if the node has any modifiers
-        if n.name in node_modifiers:
-            # if node is an output, then modify the node to be a rectangle
-            if node_modifiers[n.name].get("is_output"):
-                other_args["shape"] = "rectangle"
-        digraph.node(n.name, label=f"Input: {n.name}", **other_args)
-
-    for n in list(nodes) + list(user_nodes):
+            if node_modifiers[n.name].get("is_path"):
+                other_args["color"] = "red"
```
Comment (elijahbenizzy) [inline hamilton/graph.py:126]: Make a constant maybe?

File: hamilton/graph.py [hunk 4]
```diff
@@ -95,43 +95,51 @@ def create_function_graph(
...
+            if strictly_display_only_nodes_passed_in and d not in nodes:
+                continue
+            if (
+                d not in nodes
+                and d.name in node_modifiers
+                and node_modifiers[d.name].get("is_user_input")
+            ):
+                digraph.node(d.name, label=f"Input: {d.name}", style="dashed")
+            # print(f"Adding edge from {d.name} to {n.name}")
             digraph.edge(d.name, n.name)
```
Comment (elijahbenizzy) [inline hamilton/graph.py:143]: OK, should be easy to make the edges colored as well -- would be nice.

```python
extra_edge_args = {"color" : PATH_COLOR} if node_modifiers['d.name'].get('is_user_input') and ... else
```

File: hamilton/graph.py [hunk 5]
```diff
@@ -95,43 +95,51 @@ def create_function_graph(
...
         for d in n.dependencies:
+            if strictly_display_only_nodes_passed_in and d not in nodes:
+                continue
+            if (
+                d not in nodes
+                and d.name in node_modifiers
+                and node_modifiers[d.name].get("is_user_input")
+            ):
+                digraph.node(d.name, label=f"Input: {d.name}", style="dashed")
+            # print(f"Adding edge from {d.name} to {n.name}")
```
Comment (elijahbenizzy) [inline hamilton/graph.py:142]: delete

File: hamilton/graph.py [hunk 6]
```diff
@@ -211,23 +219,22 @@ def display_all(
...
-        defined_nodes = set()
-        user_nodes = set()
+        all_nodes = set()
+        node_modifiers = {}
         for n in self.nodes.values():
             if n.user_defined:
-                user_nodes.add(n)
-            else:
-                defined_nodes.add(n)
+                node_modifiers[n.name] = {"is_user_input": True}
```
Comment (elijahbenizzy) [inline hamilton/graph.py:226]: constants would be nice for these as we use the same value twice. An object/dataclass would work but I think this is overall pretty clean.

**Review by elijahbenizzy — COMMENTED** [review]
LGTM -- some small nits

File: hamilton/graph.py (hunk 3 above)
Comment (skrawcz) [reply hamilton/graph.py:126]: it's not used anywhere else - if it was -- then yes.

File: hamilton/graph.py (hunk 4 above)
Comment (skrawcz) [reply hamilton/graph.py:143]: feature creep. maybe later.

File: hamilton/graph.py (hunk 6 above)
Comment (skrawcz) [reply hamilton/graph.py:226]: created enum.

**Pull request merged**